crates/bevy_text/src/pipeline.rs

```diff
@@ -1,52 +1,147 @@
-use crate::{
-    compute_text_bounds, error::TextError, glyph_brush::GlyphBrush, scale_value, BreakLineOn, Font,
-    FontAtlasSets, FontAtlasWarning, PositionedGlyph, Text, TextAlignment, TextSection,
-    TextSettings, YAxisOrientation,
-};
-use ab_glyph::PxScale;
-use bevy_asset::{AssetId, Assets, Handle};
-use bevy_ecs::component::Component;
-use bevy_ecs::prelude::ReflectComponent;
-use bevy_ecs::system::Resource;
+use std::sync::{Arc, Mutex};
+
+use bevy_asset::{AssetId, Assets};
+use bevy_ecs::{component::Component, reflect::ReflectComponent, system::Resource};
 use bevy_math::Vec2;
-use bevy_reflect::prelude::ReflectDefault;
-use bevy_reflect::Reflect;
+use bevy_reflect::{std_traits::ReflectDefault, Reflect};
 use bevy_render::texture::Image;
 use bevy_sprite::TextureAtlas;
-use bevy_utils::HashMap;
-use glyph_brush_layout::{FontId, GlyphPositioner, SectionGeometry, SectionText, ToSectionText};
+use bevy_utils::{tracing::warn, HashMap};
 
-#[derive(Default, Resource)]
-pub struct TextPipeline {
-    brush: GlyphBrush,
-    map_font_id: HashMap<AssetId<Font>, FontId>,
+use cosmic_text::{Attrs, Buffer, Metrics, Shaping, Wrap};
+
+use crate::{
+    error::TextError, BreakLineOn, Font, FontAtlasSets, FontAtlasWarning, PositionedGlyph,
+    TextAlignment, TextSection, TextSettings, YAxisOrientation,
+};
+
+// TODO: cache buffers / store buffers on the entity
```

Comment:
Not caching buffers seems like it could potentially be quite a big performance issue. I'll need to read the full code to see the state of things, but if they're not cached at all I believe that would mean that shaping (expensive) happens every frame instead of only when text changes.

Reply:
If we can cache this, then exposing advanced shaping (probably by default) seems like the right call.